### PR TITLE
nfs: restart nfs servers when configmap is updated

### LINKED ISSUE #9028: Automatically restart NFS-Ganesha container when the ConfigMap is updated
**Is this a bug report or feature request?**
* Feature Request

If there is an update to a CephNFS resource, Rook updates a `rook-ceph-nfs-<name>-<id>` configmap. Rook updates the configmap, but there is usually no change to the deployment. This means the `ganesha.nfsd` process isn't restarted, and the config change doesn't take effect.

There are a few ways to implement this behavior:
1. Rook could delete the pod in order to restart the process if the configmap is updated.
2. Rook could add a sidecar container that watches the file where the configmap is mounted and issues a restart to the `ganesha.nfsd` process. e.g., https://william-yeh.net/post/2019/06/autoreload-from-configmap/
3. Get a hash of the configmap and apply it as an annotation to the deployment. e.g., https://blog.questionable.services/article/kubernetes-deployments-configmap-change/
4. possibly others

Of the explored options here, number 3 is probably the best. Number 1 could be prone to issues if an NFS deployment is currently scaling down one pod and up another. Number 2 requires a sidecar, and there are possibilities of bugs in the restart sidecar process, whether from inotify or some other app, and it's also necessary to mount the containers such that one can restart the other. Number 3 is straightforward, and while bespoke, can be unit tested easily.

### PR DESCRIPTION
**Pull request by BlaineEXE**
When the configuration configmap is updated for a CephNFS server, the
NFS application should restart to ensure it is running with the latest
config.

Fixes #9028

Signed-off-by: Blaine Gardner <blaine.gardner@redhat.com>

**Description of your changes:**

**Which issue is resolved by this Pull Request:**
Resolves #

**Checklist:**

- [ ] **Commit Message Formatting**: Commit titles and messages follow guidelines in the [developer guide](https://rook.io/docs/rook/latest/development-flow.html#commit-structure).
- [ ] **Skip Tests for Docs**: Add the flag for skipping the build if this is only a documentation change. See [here](https://github.com/rook/rook/blob/master/INSTALL.md#skip-ci) for the flag.
- [ ] **Skip Unrelated Tests**: Add a flag to run tests for a specific storage provider. See [test options](https://github.com/rook/rook/blob/master/INSTALL.md#test-storage-provider).
- [ ] Reviewed the developer guide on [Submitting a Pull Request](https://rook.io/docs/rook/latest/development-flow.html#submitting-a-pull-request)
- [ ] Documentation has been updated, if necessary.
- [ ] Unit tests have been added, if necessary.
- [ ] Integration tests have been added, if necessary.
- [ ] Pending release notes updated with breaking and/or notable changes, if necessary.
- [ ] Upgrade from previous release is tested and upgrade user guide is updated, if necessary.
- [ ] Code generation (`make codegen`) has been run to update object specifications, if necessary.

### CODE REVIEW

File: pkg/operator/ceph/nfs/nfs_test.go [hunk 1]
```diff
@@ -0,0 +1,209 @@
...
+		names = append(names, dep.Name)
+		assert.Contains(t, dep.Spec.Template.Annotations, "config-hash")
+		hashes = append(hashes, dep.Spec.Template.Annotations["config-hash"])
+	}
+	assert.ElementsMatch(t, []string{"rook-ceph-nfs-my-nfs-a", "rook-ceph-nfs-my-nfs-b"}, names)
+	assert.NotEqual(t, hashes[0], hashes[1])
+
+	svcs, err := r.context.Clientset.CoreV1().Services(ns).List(context.TODO(), metav1.ListOptions{})
+	assert.NoError(t, err)
+	// Each NFS server gets a service.
```
Comment (BlaineEXE) [inline pkg/operator/ceph/nfs/nfs_test.go:202]: This is the current behavior, but I'm not sure this is right. Should this be the design?

File: pkg/operator/ceph/nfs/nfs.go [hunk 2]
```diff
@@ -191,28 +194,33 @@ func (r *ReconcileCephNFS) generateConfigMap(n *cephv1.CephNFS, name string) *v1
...
 		logger.Debugf("updating config map %q that already exists", configMap.Name)
 		if _, err = r.context.Clientset.CoreV1().ConfigMaps(n.Namespace).Update(r.opManagerContext, configMap, metav1.UpdateOptions{}); err != nil {
-			return "", errors.Wrap(err, "failed to update ganesha config map")
+			return "", "", errors.Wrap(err, "failed to update ganesha config map")
 		}
 	}
 
-	return configMap.Name, nil
+	h := sha256.New()
+	h.Write([]byte(fmt.Sprintf("%v", configMap.Data)))
```
Comment (leseb) [inline pkg/operator/ceph/nfs/nfs.go:220]: No error handling?

**Review by leseb — CHANGES_REQUESTED** [review]

File: pkg/operator/ceph/nfs/nfs_test.go (hunk 1 above)
Comment (BlaineEXE) [reply pkg/operator/ceph/nfs/nfs_test.go:202]: This is part of the design doc: 

> The operator creates a k8s service for each of the ganesha server pods
  to allow each of the them to have a stable IP address.

File: pkg/operator/ceph/nfs/spec.go [hunk 3]
```diff
@@ -141,6 +141,12 @@ func (r *ReconcileCephNFS) makeDeployment(nfs *cephv1.CephNFS, cfg daemonConfig)
 		ObjectMeta: metav1.ObjectMeta{
 			Name:   resourceName,
 			Labels: getLabels(nfs, cfg.ID, true),
+			Annotations: map[string]string{
```
Comment (travisn) [inline pkg/operator/ceph/nfs/spec.go:144]: You're seeing an update to the annotations cause a pod restart? I was thinking the annotations didn't affect the pod restart, but that must only be the case when the annotations are on the deployment instead of the podTemplateSpec.

File: pkg/operator/ceph/nfs/nfs.go [hunk 4]
```diff
@@ -46,17 +47,18 @@ const (
 var updateDeploymentAndWait = opmon.UpdateCephDeploymentAndWait
 
 type daemonConfig struct {
-	ID              string              // letter ID of daemon (e.g., a, b, c, ...)
-	ConfigConfigMap string              // name of configmap holding config
-	DataPathMap     *config.DataPathMap // location to store data in container
+	ID                  string              // letter ID of daemon (e.g., a, b, c, ...)
+	ConfigConfigMap     string              // name of configmap holding config
```
Comment (travisn) [inline pkg/operator/ceph/nfs/nfs.go:51]: How is this configmap related to the RADOS object where the config is stored? If the RADOS config is updated, do we get events for that to restart the daemon as well?

File: pkg/operator/ceph/nfs/spec.go (hunk 3 above)
Comment (leseb) [reply pkg/operator/ceph/nfs/spec.go:144]: I thought I had a similar comment in my initial review, somehow I cannot find it anymore. Why not use the deployment annotation instead of the pod's one?

Comment (BlaineEXE) [reply pkg/operator/ceph/nfs/spec.go:144]: Because a changed annotation on the deployment won't cause the pod to restart, but a changed annotation on the pod should.

File: pkg/operator/ceph/nfs/nfs.go (hunk 4 above)
Comment (BlaineEXE) [reply pkg/operator/ceph/nfs/nfs.go:51]: The RADOS config contains additional config. This config is needed in order for Ganesha to find the RADOS config(s).

**Comment by BlaineEXE** [comment]
FWIW, this provides some idea of why config reloading is necessary, though it is from 2017: https://github.com/nfs-ganesha/nfs-ganesha/issues/173

File: pkg/operator/ceph/nfs/spec.go (hunk 3 above)
Comment (BlaineEXE) [reply pkg/operator/ceph/nfs/spec.go:144]: Confirmed that the annotation causes a pod restart as desired.

File: pkg/operator/ceph/nfs/nfs.go (hunk 4 above)
Comment (travisn) [reply pkg/operator/ceph/nfs/nfs.go:51]: Is this something worth documenting? Or maybe an overhaul of the [ceph nfs](https://rook.github.io/docs/rook/latest/ceph-nfs-crd.html) doc is better left for a separate PR.

Comment (BlaineEXE) [reply pkg/operator/ceph/nfs/nfs.go:51]: This isn't something that users should be touching, so I don't think it's worth documenting. It's what allows us to run the NFS servers with a generated config that doesn't get made in an init container that needs to run the Rook binary.

Comment (travisn) [reply pkg/operator/ceph/nfs/nfs.go:51]: one more question... When is the configmap updated? When exports are added/removed?

**Review by travisn — APPROVED** [review]

File: pkg/operator/ceph/nfs/nfs.go (hunk 4 above)
Comment (leseb) [reply pkg/operator/ceph/nfs/nfs.go:51]: Only when the CephNFS CR is updated AFAIK.

File: pkg/operator/ceph/nfs/spec.go (hunk 3 above)
Comment (leseb) [reply pkg/operator/ceph/nfs/spec.go:144]: One more question, the configmap is updated during the reconcile right? So regardless of the annotation being on pod or the deployment, the pod is going to be restarted right? If this is true I don't understand why we don't place the annotation on the deployment instead.
Quickly glancing at the code, I doubt the CephNFS config should be changed or supports any changes despite `GaneshaServerSpec`. The only CR spec that impacts the configmap is `n.Spec.RADOS`, especially `n.Spec.RADOS.Namespace`, is it safe to change it? Since recently changing `n.Spec.RADOS.Pool` does not do anything since it always defaults to `.nfs`.
If all the above is true and changing `n.Spec.RADOS.Namespace` is risky then I'm wondering why we would need to restart the pod if the config changes if there is a misconfiguration in the CR perhaps? I just don't see where it could be.
Thanks.

**Review by leseb — CHANGES_REQUESTED** [review]
I still need some more clarification. Thanks

**Comment by BlaineEXE** [comment]
> One more question, the configmap is updated during the reconcile right? So regardless of the annotation being on pod or the deployment, the pod is going to be restarted right? If this is true I don't understand why we don't place the annotation on the deployment instead.

This is during reconcile, but if they pod template spec in the deployment doesn't change, then the pod doesn't get restarted, even if the deployment itself changes. That is why the annotation is needed on the pod.

> Quickly glancing at the code, I doubt the CephNFS config should be changed or supports any changes despite GaneshaServerSpec. The only CR spec that impacts the configmap is n.Spec.RADOS, especially n.Spec.RADOS.Namespace, is it safe to change it? Since recently changing n.Spec.RADOS.Pool does not do anything since it always defaults to .nfs.
If all the above is true and changing n.Spec.RADOS.Namespace is risky then I'm wondering why we would need to restart the pod if the config changes if there is a misconfiguration in the CR perhaps? I just don't see where it could be.

This is all also correct. These updates are only useful on Ceph Octopus, where the config is still allowed to change. I was thinking that this would be desirable for migrating to the `.nfs` pool **before** upgrading from Octopus to Pacific. It could allow the upgrade to be without downtime instead of with a ?-amount of downtime.

**Comment by leseb** [comment]
> > One more question, the configmap is updated during the reconcile right? So regardless of the annotation being on pod or the deployment, the pod is going to be restarted right? If this is true I don't understand why we don't place the annotation on the deployment instead.
> 
> This is during reconcile, but if they pod template spec in the deployment doesn't change, then the pod doesn't get restarted, even if the deployment itself changes. That is why the annotation is needed on the pod.
> 
> > Quickly glancing at the code, I doubt the CephNFS config should be changed or supports any changes despite GaneshaServerSpec. The only CR spec that impacts the configmap is n.Spec.RADOS, especially n.Spec.RADOS.Namespace, is it safe to change it? Since recently changing n.Spec.RADOS.Pool does not do anything since it always defaults to .nfs.
> > If all the above is true and changing n.Spec.RADOS.Namespace is risky then I'm wondering why we would need to restart the pod if the config changes if there is a misconfiguration in the CR perhaps? I just don't see where it could be.
> 
> This is all also correct. These updates are only useful on Ceph Octopus, where the config is still allowed to change. I was thinking that this would be desirable for migrating to the `.nfs` pool **before** upgrading from Octopus to Pacific. It could allow the upgrade to be without downtime instead of with a ?-amount of downtime.

Sorry earlier I meant `GaneshaRADOSSpec` but the underlying specs were correct (`RADOS.Namespace`).
Understood for the migration, we need to migrate data from the old pool to `.nfs` and then change `n.Spec.RADOS.Pool` to `.nfs`. Then with this patch, the pod will restart so that the FSAL can be updated. I think this last comment is good for posterity, now we know the WHY this patch is needed.

**Review by leseb — APPROVED** [review]
Finally approving based on https://github.com/rook/rook/pull/9104#issuecomment-980079136. Thanks!
